File: Darts/src/burlton/dartzee/code/screen/stats/player/MovingAverageChartPanel.java
package burlton.dartzee.code.screen.stats.player;

import burlton.dartzee.code.utils.DartsColour;
import burlton.desktopcore.code.util.ComponentUtilKt;
import org.jfree.chart.ChartFactory;
import org.jfree.chart.ChartPanel;
import org.jfree.chart.JFreeChart;
import org.jfree.chart.plot.XYPlot;
import org.jfree.data.time.MovingAverage;
import org.jfree.data.xy.XYSeries;
import org.jfree.data.xy.XYSeriesCollection;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.List;

public class MovingAverageChartPanel extends JPanel
									 implements ActionListener
{
	private XYSeriesCollection graphCollection = new XYSeriesCollection();
	private JFreeChart chart = null;
	private AbstractStatisticsTab parentTab = null;
	
	public MovingAverageChartPanel(AbstractStatisticsTab parentTab) 
	{
		this.parentTab = parentTab;
		
		setLayout(new BorderLayout(0, 0));
		panelGraph.setLayout(new BorderLayout(0, 0));
		
		add(panelGraph, BorderLayout.CENTER);
		add(panelCheckBoxes, BorderLayout.SOUTH);
	}
	
	private final ChartPanel panelGraph = new ChartPanel(null);
	private final JPanel panelCheckBoxes = new JPanel();
	
	public void init(String title, String yLabel)
	{
		graphCollection = new XYSeriesCollection();
		
		chart = ChartFactory.createXYLineChart(title, "Game", yLabel, graphCollection);
		panelGraph.setChart(chart);
	}
	
	public void finalise()
	{
		adjustTickboxes();
		setPlotColours(chart);
	}
	
	private void adjustTickboxes()
	{
		List<JCheckBox> checkBoxes = ComponentUtilKt.getAllChildComponentsForType(panelCheckBoxes, JCheckBox.class);
		
		//Remove any bad ones
		for (JCheckBox checkBox : checkBoxes)
		{
			if (getGraphSeriesIndexForCheckBox(checkBox) == -1)
			{
				panelCheckBoxes.remove(checkBox);
			}
		}
		
		//Update this variable
		checkBoxes = ComponentUtilKt.getAllChildComponentsForType(panelCheckBoxes, JCheckBox.class);
		
		//Go through each series in the graph and ensure it has a checkbox
		List<XYSeries> allSeries = graphCollection.getSeries();
		for (int i=0; i<allSeries.size(); i++)
		{
			XYSeries series = allSeries.get(i);
			String key = (String)series.getKey();
			if (matchingCheckBoxExists(checkBoxes, key))
			{
				continue;
			}
			
			JCheckBox checkBox = new JCheckBox(key);
			checkBox.setSelected(true);
			checkBox.addActionListener(this);
			panelCheckBoxes.add(checkBox);
		}
		
		//Now for any checkBox that remains and is unticked, remove the graph series
		for (JCheckBox checkBox : checkBoxes)
		{
			if (checkBox.isSelected())
			{
				continue;
			}
			
			int graphSeries = getGraphSeriesIndexForCheckBox(checkBox);
			graphCollection.removeSeries(graphSeries);
		}
	}
	private boolean matchingCheckBoxExists(List<JCheckBox> checkBoxes, String graphKey)
	{
		for (JCheckBox checkBox : checkBoxes)
		{
			String cbText = checkBox.getText();
			if (cbText.equals(graphKey))
			{
				return true;
			}
		}
		
		return false;
	}
	private int getGraphSeriesIndexForCheckBox(JCheckBox checkBox)
	{
		String text = checkBox.getText();
		
		List<XYSeries> allSeries = graphCollection.getSeries();
		for (int i=0; i<allSeries.size(); i++)
		{
			XYSeries series = allSeries.get(i);
			String key = (String)series.getKey();
			
			if (key.equals(text))
			{
				return i;
			}
		}
		
		return -1;
	}
	
	private void setPlotColours(JFreeChart chart)
	{
		XYPlot plot = (XYPlot)chart.getPlot();
		
		List<XYSeries> allSeries = graphCollection.getSeries();
		for (int i=0; i<allSeries.size(); i++)
		{
			XYSeries series = allSeries.get(i);
			String key = (String)series.getKey();
			
			Color colour = Color.blue;
			if (key.contains("Other"))
			{
				colour = Color.red;
			}
			
			if (key.contains("Moving"))
			{
				colour = DartsColour.getBrightenedColour(colour);
			}
			
			plot.getRenderer().setSeriesPaint(i, colour);
		}
	}
	
	public void addSeries(XYSeries series, String key, int avgThreshold)
	{
		XYSeries movingAvgSeries = createMovingAverage(series, key, avgThreshold);
		graphCollection.addSeries(series);
		graphCollection.addSeries(movingAvgSeries);
	}
	private XYSeries createMovingAverage(XYSeries original, String key, int avgThreshold)
	{
		XYSeriesCollection collection = new XYSeriesCollection(original);
		XYSeriesCollection movingAvgCollection = (XYSeriesCollection)MovingAverage.createMovingAverage(collection, "", avgThreshold, avgThreshold-1);
		
		XYSeries movingAvgSeries = movingAvgCollection.getSeries(0);
		movingAvgSeries.setKey("Moving avg " + key);
		return movingAvgSeries;
	}
	
	@Override
	public void actionPerformed(ActionEvent arg0)
	{
		parentTab.populateStats();
	}
}

File: Darts/src/burlton/dartzee/code/screen/stats/player/MovingAverageChartPanel.kt
package burlton.dartzee.code.screen.stats.player

import burlton.dartzee.code.utils.DartsColour.getBrightenedColour
import burlton.dartzee.code.core.util.getAllChildComponentsForType
import org.jfree.chart.ChartFactory
import org.jfree.chart.ChartPanel
import org.jfree.chart.JFreeChart
import org.jfree.chart.plot.XYPlot
import org.jfree.data.time.MovingAverage
import org.jfree.data.xy.XYSeries
import org.jfree.data.xy.XYSeriesCollection
import java.awt.BorderLayout
import java.awt.Color
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import javax.swing.JCheckBox
import javax.swing.JPanel

class MovingAverageChartPanel(private val parentTab: AbstractStatisticsTab) : JPanel(), ActionListener
{
    private var graphCollection = XYSeriesCollection()
    private val panelGraph = ChartPanel(null)
    private val panelCheckBoxes = JPanel()

    init
    {
        layout = BorderLayout(0, 0)
        panelGraph.layout = BorderLayout(0, 0)
        add(panelGraph, BorderLayout.CENTER)
        add(panelCheckBoxes, BorderLayout.SOUTH)
    }

    fun reset(title: String, yLabel: String)
    {
        graphCollection = XYSeriesCollection()
        panelGraph.chart = ChartFactory.createXYLineChart(title, "Game", yLabel, graphCollection)
    }

    fun finalise()
    {
        adjustTickboxes()
        setPlotColours(panelGraph.chart)
    }

    private fun adjustTickboxes()
    {
        //Remove checkboxes that are no longer relevant (e.g. just removed a comparison)
        getAllChildComponentsForType(panelCheckBoxes, JCheckBox::class.java).forEach {
            if (getGraphSeriesIndexForCheckBox(it) == -1) panelCheckBoxes.remove(it)
        }

        //Go through what's left and add any others that are required by the graph (e.g. just added a comparison)
        val checkBoxes = getAllChildComponentsForType(panelCheckBoxes, JCheckBox::class.java)
        val allSeries: List<XYSeries> = graphCollection.getXYSeries()
        allSeries.forEach {
            val key = "${it.key}"
            if (checkBoxes.none { cb -> cb.text == key} )
            {
                val checkBox = JCheckBox(key)
                checkBox.isSelected = true
                checkBox.addActionListener(this)
                panelCheckBoxes.add(checkBox)
            }
        }

        //For any checkboxes that are unticked, remove the graph series
        checkBoxes.filter { !it.isSelected }.forEach { graphCollection.removeSeries(getGraphSeriesIndexForCheckBox(it)) }

        panelCheckBoxes.validate()
        panelCheckBoxes.repaint()
    }

    private fun getGraphSeriesIndexForCheckBox(checkBox: JCheckBox): Int
    {
        val allSeries: List<XYSeries> = graphCollection.getXYSeries()
        return allSeries.indexOfFirst { it.key == checkBox.text }
    }

    private fun XYSeriesCollection.getXYSeries() = series.filterIsInstance<XYSeries>()

    private fun setPlotColours(chart: JFreeChart?)
    {
        val plot = chart!!.plot as XYPlot
        val allSeries: List<XYSeries> = graphCollection.getXYSeries()

        allSeries.forEachIndexed { ix, series ->
            val key = "${series.key}"
            var colour = Color.blue
            if (key.contains("Other")) {
                colour = Color.red
            }
            if (key.contains("Moving")) {
                colour = getBrightenedColour(colour!!)
            }
            plot.renderer.setSeriesPaint(ix, colour)
        }
    }

    fun addSeries(series: XYSeries, key: String, avgThreshold: Int)
    {
        val movingAvgSeries = createMovingAverage(series, key, avgThreshold.toLong())
        graphCollection.addSeries(series)
        graphCollection.addSeries(movingAvgSeries)
    }

    private fun createMovingAverage(original: XYSeries, key: String, avgThreshold: Long): XYSeries
    {
        val collection = XYSeriesCollection(original)
        val movingAvgCollection = MovingAverage.createMovingAverage(collection,"", avgThreshold,avgThreshold - 1)
        val movingAvgSeries = (movingAvgCollection as XYSeriesCollection).getSeries(0)
        movingAvgSeries.key = "Moving avg $key"
        return movingAvgSeries
    }

    override fun actionPerformed(arg0: ActionEvent)
    {
        parentTab.populateStats()
    }
}